design/baremetal-operator/firmware-image-update.md

```diff
@@ -0,0 +1,195 @@
+      checksumType: sha1
+      vendor: nvidia
+      component: net
+      componentFlavor: MT_0000000652
+      version: 22.35.1012
+```
+
+The `HostFirmwareImage` status section will introduce the following attributes:
+
+- Firmwares – list of all hardware devices with the following information:
+    - componentID – unique identifier for the hardware. For NICs the unique identifier will be in the following format “`<interface name>_<MAC>”.`
+    - vendor – The hardware vendor which is responsible for that firmware.
+    - component – The hardware component type.
+    - componentFlavor - indicate the component flavor if such exist.
+    - version - the actual firmware version of the hardware.
+- lastUpdated – The timestamp of the last update.
+- Conditions – we introduce 2 conditions:
+    - FirmwareImageChangeDetected - Indicates that the firmware images version in the spec are different than status.
+    - FirmwareImageValid - Indicates that firmware images are valid and can be configured on the host.
+
+```yaml
+status:
+  firmwares:
+    - componentID: enp216s0f0np0_08_c0_eb_70_74_62
+      vendor: nvidia
+      component: net
+      componentFlavor: MT_1090120019
+      version: 24.34.1002
+    - componentID: enp216s0f1np1_08_c0_eb_70_74_63
+      vendor: nvidia
+      component: net
+      componentFlavor: MT_1090120019
+      version: 24.34.1002
+    - componentID: enp59s0f0np0_b8_ce_f6_8d_c8_62
+      vendor: nvidia
+      component: net
+      componentFlavor: MT_0000000652
+      version: 22.35.1012
+  lastUpdated: 2022-12-27T01:12:17Z
+  conditions: ...
+```
+
+### HostFirmwareSettings Controller Changes
+
+The proposal is to extend the `HostFirmwareSettings` controller which was introduced in the `bulk-set-bios-config` proposal. The controller will be extended to track the `HostFirmwareImage` CRD. The HostFirmwareImage.spec is the desired firmware version and the HostFirmwareImage.status contain the actual firmware. Please note that the correct proposal addresses Nvidia NICs but can be extended to other vendors too as well.
+
+The NIC firmware version exists today in the data.Extra.Network struct. We will also extend `HardwareData` with the NIC firmware by adding firmware attribute to the details.NIC. BIOS version already exists in `HardwareData` but other devices which do not firmware version is not exposed will required to provide mechanism to retrieve actual firmware and store it in the `HardwareData`.
+
+The reconcile loop of the `HostFirmwareSettings` controller will do the following:
+
+- Validate the `HostFirmwareImage` attributes (e.g., vendor name, component …). We will try to validate at match as we can. If valid set condition to FirmwareImageValid to true.
+- Fetch the corresponding HardwareData and get the Nvidia NIC firmware. The Nvidia firmware contains firmware version and componentFlavor "24.34.1002 (MT_0000000540)" other NIC vendors can extend the parsing according to their NIC firmware string.

```

Comment:
This requires knowledge of how to get the firmware version for a particular piece of hardware from HardwareDetails to be built in to the controller.
It's really Ironic's job to be providing the abstraction over different kinds of hardware. Ironic has third-party CI set up to run against multiple vendors' hardware. We don't have that kind of community in Metal3, so we try to avoid being in the business of having to implement the hardware abstraction.